File: Algorithms/leetcode/Ex_513_Find_Bottom_Left_Tree_Value/Soln_3_Recursive_Dfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findBottomLeftValue(self, root: TreeNode) -> int:
        if not root:
            return 0
        leftmost = []
        self.dfs(root, leftmost, 1)
        return leftmost[-1]

    def dfs(self, node: TreeNode, leftmost: List[TreeNode], depth: int) -> None:
        if not node:
            return
        if depth > len(leftmost):
            leftmost.append(node.val)
        self.dfs(node.left, leftmost, depth+1)
        self.dfs(node.right, leftmost, depth+1)


# 参考：https://leetcode-cn.com/problems/find-bottom-left-tree-value/solution/yan-du-you-xian-sou-suo-bfs-shen-du-you-lgyhj/

File: Algorithms/leetcode/Ex_0257_Binary_Tree_Paths/Soln_3_Recursive_Dfs.cpp
#include <vector>
#include <string>

using std::vector;
using std::string;
using std::to_string;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution_3 {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> res;
        if (root == nullptr) {
            return res;
        }
        dfs(root, "", res);
        return res;
    }

private:
    void dfs(TreeNode *cur, string path, vector<string> &res) {
        if (cur->left == nullptr && cur->right == nullptr) {
            path += to_string(cur->val);
            res.push_back(path);
            return;
        }

        if (cur->left) {
            // 修改当前节点状态：
            //      cur = cur->left
            //      path = path + to_string(cur->val) + "->"
            // 递归子节点
            // 每当退回到当前状态时，节点状态自动恢复
            dfs(cur->left, path + to_string(cur->val) + "->", res);
        }
        if (cur->right) {
            // 同上，修改当前节点状态、递归子节点、恢复当前节点状态 三者合并为一条语句
            dfs(cur->right, path + to_string(cur->val) + "->", res);
        }

    }
};

// // 参考「编程指北」《谷歌大佬刷题笔记》（强烈推荐）：https://zhuanlan.zhihu.com/p/297648394
